File: python/day12.py
from utilities import read_lines
from typing import List

directions = {'N': 0, 'E': 1, 'S': 2, 'W': 3}
direction_factors = [(0, 1), (1, 0), (0, -1), (-1, 0)]


def calculate_manhattan_distance(instructions: List[str]) -> int:
    # Starting position
    x, y = 0, 0

    # Starting direction is east
    direction_index = directions['E']

    for instruction in instructions:
        value = int(instruction[1:])
        action = instruction[0]
        if action in directions:
            direction = direction_factors[directions[action]]
            x += value * direction[0]
            y += value * direction[1]
        elif action == 'F':
            x += value * direction_factors[direction_index][0]
            y += value * direction_factors[direction_index][1]
        else:
            factor = -1
            if action == 'R':
                factor = 1
            direction_index += (value // 90) * factor + 4
            direction_index %= 4

    return abs(x) + abs(y)


def calculate_manhattan_distance_with_waypoint(instructions):
    # Starting position
    x, y = 0, 0
    way_x, way_y = 10, 1

    for instruction in instructions:
        value = int(instruction[1:])
        action = instruction[0]

        if action in directions:
            direction = direction_factors[directions[action]]
            way_x += value * direction[0]
            way_y += value * direction[1]
        elif action == 'F':
            x += way_x * value
            y += way_y * value
        elif value == 180:
            way_x *= -1
            way_y *= -1
        else:
            rotation = value
            if action == 'L':
                rotation = 360 - value

            factor_x = 1 if rotation == 90 else -1
            factor_y = -1 if rotation == 90 else 1

            way_x, way_y = factor_x * way_y, factor_y * way_x

    return abs(x) + abs(y)


def part_1(instructions: List[str]) -> int:
    return calculate_manhattan_distance(instructions)


def part_2(instructions: List[str]) -> int:
    return calculate_manhattan_distance_with_waypoint(instructions)


if __name__ == '__main__':
    print('Day 12')
    instructions = read_lines('../inputs/day12.txt')
    print('\tPart 1: {}'.format(part_1(instructions)))
    print('\tPart 2: {}'.format(part_2(instructions)))

File: rust/src/day12.rs
use crate::utilities::read_lines;

const DIRECTIONS: [char; 4] = ['N', 'E', 'S', 'W'];
const DIRECTION_FACTORS: [(i32, i32); 4] = [(0, 1), (1, 0), (0, -1), (-1, 0)];

fn calculate_manhattan_distance(instructions: &Vec<String>) -> i32 {
    let (mut x, mut y) = (0i32, 0i32); // Starting position

    let mut direction_index = 1usize; // Initially facing east

    for instruction in instructions {
        let action = instruction.chars().next().unwrap();
        let value: i32 = instruction.as_str()[1..].parse().unwrap();

        if action == 'F' {
            x += value * DIRECTION_FACTORS[direction_index].0;
            y += value * DIRECTION_FACTORS[direction_index].1;
        } else if action == 'R' {
            direction_index += value as usize / 90;
            direction_index %= 4;
        } else if action == 'L' {
            direction_index += 4 - value as usize / 90;
            direction_index %= 4;
        } else {
            for i in 0..4 {
                if DIRECTIONS[i] == action {
                    x += value * DIRECTION_FACTORS[i].0;
                    y += value * DIRECTION_FACTORS[i].1;
                    break;
                }
            }
        }
    }
    x.abs() + y.abs()
}

fn calculate_manhattan_distance_with_waypoint(instructions: &Vec<String>) -> i32 {
    let (mut x, mut y) = (0i32, 0i32); // Starting position
    let (mut waypoint_x, mut waypoint_y) = (10i32, 1i32); // Starting waypoint

    for instruction in instructions {
        let action = instruction.chars().next().unwrap();
        let value: i32 = instruction.as_str()[1..].parse().unwrap();

        if action == 'F' {
            x += value * waypoint_x;
            y += value * waypoint_y;
        } else if action == 'R' || action == 'L' {
            if value == 180 {
                waypoint_x = -waypoint_x;
                waypoint_y = -waypoint_y;
            } else {
                let right_rotation = if action == 'R' { value } else { 360 - value };

                let factor = if right_rotation == 90 {
                    (1, -1)
                } else {
                    (-1, 1)
                };
                let tmp = waypoint_x;
                waypoint_x = factor.0 * waypoint_y;
                waypoint_y = factor.1 * tmp;
            }
        } else {
            for i in 0..4 {
                if DIRECTIONS[i] == action {
                    waypoint_x += value * DIRECTION_FACTORS[i].0;
                    waypoint_y += value * DIRECTION_FACTORS[i].1;
                    break;
                }
            }
        }
    }
    x.abs() + y.abs()
}

fn part_1(instructions: &Vec<String>) {
    println!("\tPart 1: {}", calculate_manhattan_distance(instructions));
}

fn part_2(instructions: &Vec<String>) {
    println!(
        "\tPart 2: {}",
        calculate_manhattan_distance_with_waypoint(instructions)
    );
}

pub fn main() {
    println!("Day 12");
    let instructions = read_lines("../inputs/day12.txt");
    part_1(&instructions);
    part_2(&instructions);
}
